extensions/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/BuildProcessor.java

```diff
@@ -249,17 +262,55 @@ CamelRuntimeRegistryBuildItem bindRuntimeBeansToRegistry(
      * disabled by setting quarkus.camel.disable-main = true
      */
     public static class Main {
+
+        @BuildStep
+        public List<CamelRoutesBuilderClassBuildItem> discoverRoutesBuilderClassNames(
+                CombinedIndexBuildItem combinedIndex) {
+            final IndexView index = combinedIndex.getIndex();
+            Set<ClassInfo> allKnownImplementors = new HashSet<>();
+            allKnownImplementors.addAll(
+                    index.getAllKnownImplementors(DotName.createSimple(RoutesBuilder.class.getName())));
+            allKnownImplementors.addAll(
+                    index.getAllKnownSubclasses(DotName.createSimple(RouteBuilder.class.getName())));
+            allKnownImplementors.addAll(
+                    index.getAllKnownSubclasses(DotName.createSimple(AdviceWithRouteBuilder.class.getName())));
+
+            return allKnownImplementors
+                    .stream()
+                    // public and non-abstract
+                    .filter(ci -> ((ci.flags() & (Modifier.ABSTRACT | Modifier.PUBLIC)) == Modifier.PUBLIC))
+                    .map(ClassInfo::name)
+                    .map(CamelRoutesBuilderClassBuildItem::new)
+                    .collect(Collectors.toList());
+        }
+
         @Record(ExecutionTime.STATIC_INIT)
         @BuildStep
-        public List<CamelRoutesBuilderBuildItem> collectRoutes(
-                CombinedIndexBuildItem combinedIndex,
+        public List<CamelBeanBuildItem> collectRoutes(
+                List<CamelRoutesBuilderClassBuildItem> camelRoutesBuilders,
```

Comment:
maybe the name should reflect that fact thet it is about classes and not instances, something like `camelRoutesBuilderClasses` ?

Reply:
np